control_panel_api/models.py

```diff
@@ -206,3 +207,25 @@ class UserS3Bucket(AccessToS3Bucket):
     class Meta:
         # one record per user/s3bucket
         unique_together = ('user', 's3bucket')
+        ordering = ('id',)
+
+    def aws_create(self):
+        services.attach_bucket_access_to_role(
+            self.s3bucket.name,
+            self.has_readwrite_access(),
+            self.user.aws_role_name,
+        )
+
+    def aws_delete(self):
+        services.detach_bucket_access_from_role(
+            self.s3bucket.name,
+            self.has_readwrite_access(),
+            self.user.aws_role_name,
+        )
+
+    def aws_update(self):
+        services.update_bucket_access(
+            self.s3bucket.name,
+            self.has_readwrite_access(),
+            self.user.aws_role_name,
```

Comment:
Just an observation: Now seems like the only difference between these methods in `AppS3Bucket` and `UserS3Bucket` is the role to which the S3 bucket policies are attached/detached.

It makes me wonder if we could use template method to push these three methods up into `AccessToS3Bucket`, e.g. subclasses implements a method which returns the IAM role name...something like this:

```python
class AccessToS3Bucket(TimeStampedModel):

    def aws_role_name():
          raise NotImplementedError

    def aws_create(self):
        services.attach_bucket_access_to_role(
            self.s3bucket.name,
            self.has_readwrite_access(),
            self.aws_role_name(), # <-- Note this
        )

    # similar thing for `aws_update()`/`aws_delete()`
   

class AppS3Bucket(AccessToS3Bucket):

    def aws_role_name():
          return self.app.aws_role_name


class UserS3Bucket(AccessToS3Bucket):

    def aws_role_name():
          return self.user.aws_role_name
```

I don't know if any of this make sense @jmoz - what do you reckon?

Reply:
Yep that's good OO I'll use your code

Reply:
nice one